nixos/{syncoid,sanoid}: Improve ZFS permission delegation

etu opened the pull request:
###### Motivation for this change

- Sanoid does delegate the following permissions `snapshot,mount,destroy` on the whole pool of the dataset it wants to snapshot. But it also makes sure to unallow this. This PR addresses this to only delegate the permissions to datasets needed.
- Syncoid does delegate the following permissions `hold,send,snapshot,destroy` when sending and the following permissions `create,mount,receive,rollback` when receiving snapshots. Both of these set it to the whole pool rather than needed datasets. However, it doesn't ever run unallow so the permissions remain after execution. This PR addresses both of these points to only delegate permissions to datasets needed and make sure to drop them when done.

The tests (it's named sanoid, but it also tests syncoid) still works, and I've tested it both on my laptop and fileserver.

I've also took the liberty to reformat the files because the entire module files was badly indented. So it's easiest to read the diffs commit-by-commit.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

etu commented:
@ofborg build nixosTests.sanoid

grahamc commented on nixos/modules/services/backup/sanoid.nix:
```diff
@@ -70,116 +70,132 @@ let
+      serviceConfig = {
+        ExecStartPre = map
+          (dataset: lib.escapeShellArgs [
+            "+/run/booted-system/sw/bin/zfs"
```
:+1: might be worth adding a comment here saying it is explicitly the booted system to maintain the stable API needed by ZFS.

ju1m reviewed (changes requested):
Hi @etu, FYI I still maintain this old PR https://github.com/NixOS/nixpkgs/pull/98455 which uses another approach: using multiple systemd services. I find it more efficient to have syncoid runs in parallel (as it could be to different hosts), and clearer wrt. error reporting (eg. in Prometheus) to have different services, would you agree to rebase your syncoid-related work (the `zfs allow` on datasets instead of pools) upon this PR?

If yes, you could send me a pull request against https://github.com/ju1m/nixpkgs/tree/syncoid-split and put a review on https://github.com/NixOS/nixpkgs/pull/98455 . Sorry this would imply you dropping some code, I wish merging against Nixpkgs were not lagging almost a year like that.
Thanks anyway for spotting that too broad `zfs allow` :)

etu commented:
@ju1m Oh, wasn't aware of that PR. That looks like a lot of good work that solves some of my issues. I find it a bit funny with all the hardening but still exposing pool wide permissions :wink: 

But I'll try that out now on my systems. Then I can merge it and rebase my permission changes based on that work when it's on master. That PR by itself is a big improvement as it is.

etu committed:
nixos/syncoid: Extract datasets rather than pools

When sending or receiving datasets with the old implementation it
wouldn't matter which dataset we were sending or receiving, we would
always delegate permissions to the entire pool.

etu committed:
nixos/sanoid: Extract datasets rather than pools

When making new snapshots we only need to delegate permissions to the
specific dataset rather than the entire pool.

etu committed:
nixos/syncoid: Build unallow commands as a post job to drop permissions

etu committed:
nixos/sanoid: Use a function to build allow/unallow commands

etu committed:
nixos/syncoid: Reformat file with nixpkgs-fmt

etu committed:
nixos/sanoid: Reformat file with nixpkgs-fmt

etu commented:
Ping @ju1m, hey, I've "rebased" (rather, kinda, re-implemented) my changes upon master now to reduce the amounts of permissions delegated :)

ju1m commented:
Thanks @etu, here's a few remarks:
1. Maybe there should be a release entry suggesting old users of `syncoid` to **manually remove lingering permissions** to the `syncoid` user on pools (since the new `zfs unallow` will only run on datasets).
2. I guess `sanoid`'s `ExecStartPre=` (and `ExecStopPost=`) **should be prefixed by a `-`** (along the `+`) to run `sanoid` even if some datasets aren't accessible (eg. if using `sanoid` on pools which aren't always imported; `sanoid` will skip those and emit `cannot open $dataset` or `CRITICAL ERROR: zfs snapshot [...]` log messages but still return successfully; AFAICS there is no option to report via errno failure of snapshotting by `sanoid`).
3. **`~@sync` must be removed** from `syncoid`'s `SystemCallFilter=` befause `mbuffer` tries to `fsync()`. Previously I made the mistake to use`SystemCallErrorNumber=EPERM` and it was hiding the fact that `mbuffer` tries to `fsync()`, but now `ESYS` is raised and `mbuffer` dumps core instead of silently skipping the `fsync()`.

etu committed:
nixos/syncoid: Drop ~[at]sync from the systemcallfilter to avoid coredumps

etu committed:
nixos/rl-notes/21.11: Add note about remaining syncoid permissions

etu committed:
nixos/tests/sanoid: Improve tests by checking that no permissions are left behind

etu committed:
sanoid: Passthru sanoid tests

etu commented:
@ju1m I've addressed your remarks, I've also rebased everything together as I want it to look now.

I've also added checks in the tests to make sure that we don't leave any permissions behind after execution of units. Then I also added the tests to the package so it's easy to run the tests by building `syncoid.tests` :slightly_smiling_face: 

I'm currently running these syncoid/sanoid services on all my systems, seems fine so far :slightly_smiling_face:

ju1m commented:
@etu, not convinced to add `-` to `sanoid`'s `ExecStartPre=`?

etu commented:
@ju1m You mean like that?

ju1m commented:
> @ju1m You mean like that?

Yes. It was important for `sanoid` because it's a single service for all datasets. For `syncoid` that's ok too I guess, instead of `zfs allow` failing early, `syncoid` will just fail later.

etu commented:
@ju1m Now I've squashed in those changes :slightly_smiling_face:

ju1m reviewed (approved):
@etu, thanks for this work! LGTM

misuzu commented:
On nixos-unstable syncoid is able to send only one snapshot per run: 
```
Aug 07 02:00:00 myhost systemd[1]: Starting Syncoid ZFS synchronization from rpool/safe to zfs@backuphost:bpool/backup/myhost/safe...
Aug 07 02:00:00 myhost systemd[1]: Started Syncoid ZFS synchronization from rpool/safe to zfs@backuphost:bpool/backup/myhost/safe.
Aug 07 02:00:00 myhost syncoid[3207057]: Warning: Permanently added 'backuphost' (ED25519) to the list of known hosts.
Aug 07 02:00:03 myhost syncoid[3207049]: NEWEST SNAPSHOT: autosnap_2021-08-06_21:00:00_daily
Aug 07 02:00:04 myhost syncoid[3207049]: Sending incremental rpool/safe@autosnap_2021-08-05_21:00:00_daily ... autosnap_2021-08-06_21:00:00_daily (~ 693.2 MB):
Aug 07 02:00:44 myhost syncoid[3207049]: CRITICAL ERROR:  zfs send -w -R  -I 'rpool/safe'@'autosnap_2021-08-05_21:00:00_daily' 'rpool/safe'@'autosnap_2021-08-06_21:00:00_daily' | pv -p -t -e -r -b -s 726925136 | lzop  | mbuffer  -q -s 128k -m 16M 2>/dev/null | ssh  -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null  -i /run/secrets/zfs/id_ed25519 -S /tmp/syncoid-zfs@backuphost-1628290800 zfs@backuphost ' mbuffer  -q -s 128k -m 16M 2>/dev/null | lzop -dfc |  zfs receive  -s -F '"'"'bpool/backup/myhost/safe'"'"' 2>&1' failed: 256 at /nix/store/ygx9z7xqky03cvymfsja24xmmf8yd7hy-sanoid-2.1.0/bin/.syncoid-wrapped line 817.
Aug 07 02:00:44 myhost systemd[1]: syncoid-rpool-safe.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Aug 07 02:00:44 myhost systemd[1]: syncoid-rpool-safe.service: Failed with result 'exit-code'.
Aug 07 02:00:44 myhost systemd[1]: syncoid-rpool-safe.service: Consumed 12.627s CPU time, received 9.0M IP traffic, sent 655.8M IP traffic.
Aug 08 02:00:00 myhost systemd[1]: Starting Syncoid ZFS synchronization from rpool/safe to zfs@backuphost:bpool/backup/myhost/safe...
Aug 08 02:00:00 myhost systemd[1]: Started Syncoid ZFS synchronization from rpool/safe to zfs@backuphost:bpool/backup/myhost/safe.
Aug 08 02:00:00 myhost syncoid[3941498]: Warning: Permanently added 'backuphost' (ED25519) to the list of known hosts.
Aug 08 02:00:04 myhost syncoid[3941485]: NEWEST SNAPSHOT: autosnap_2021-08-07_21:00:00_daily
Aug 08 02:00:04 myhost syncoid[3941485]: Sending incremental rpool/safe@autosnap_2021-08-06_21:00:00_daily ... autosnap_2021-08-07_21:00:00_daily (~ 508.3 MB):
Aug 08 02:00:25 myhost syncoid[3941485]: CRITICAL ERROR:  zfs send -w -R  -I 'rpool/safe'@'autosnap_2021-08-06_21:00:00_daily' 'rpool/safe'@'autosnap_2021-08-07_21:00:00_daily' | pv -p -t -e -r -b -s 533040568 | lzop  | mbuffer  -q -s 128k -m 16M 2>/dev/null | ssh  -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null  -i /run/secrets/zfs/id_ed25519 -S /tmp/syncoid-zfs@backuphost-1628377200 zfs@backuphost ' mbuffer  -q -s 128k -m 16M 2>/dev/null | lzop -dfc |  zfs receive  -s -F '"'"'bpool/backup/myhost/safe'"'"' 2>&1' failed: 256 at /nix/store/ygx9z7xqky03cvymfsja24xmmf8yd7hy-sanoid-2.1.0/bin/.syncoid-wrapped line 817.
Aug 08 02:00:25 myhost systemd[1]: syncoid-rpool-safe.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Aug 08 02:00:26 myhost systemd[1]: syncoid-rpool-safe.service: Failed with result 'exit-code'.
Aug 08 02:00:26 myhost systemd[1]: syncoid-rpool-safe.service: Consumed 8.974s CPU time, received 6.3M IP traffic, sent 501.0M IP traffic.
Aug 09 02:00:00 myhost systemd[1]: Starting Syncoid ZFS synchronization from rpool/safe to zfs@backuphost:bpool/backup/myhost/safe...
Aug 09 02:00:00 myhost systemd[1]: Started Syncoid ZFS synchronization from rpool/safe to zfs@backuphost:bpool/backup/myhost/safe.
Aug 09 02:00:00 myhost syncoid[842177]: Warning: Permanently added 'backuphost' (ED25519) to the list of known hosts.
Aug 09 02:00:04 myhost syncoid[842169]: NEWEST SNAPSHOT: autosnap_2021-08-08_21:00:00_daily
Aug 09 02:00:04 myhost syncoid[842169]: Sending incremental rpool/safe@autosnap_2021-08-07_21:00:00_daily ... autosnap_2021-08-08_21:00:00_daily (~ 478.2 MB):
Aug 09 02:00:31 myhost syncoid[842169]: CRITICAL ERROR:  zfs send -w -R  -I 'rpool/safe'@'autosnap_2021-08-07_21:00:00_daily' 'rpool/safe'@'autosnap_2021-08-08_21:00:00_daily' | pv -p -t -e -r -b -s 501397960 | lzop  | mbuffer  -q -s 128k -m 16M 2>/dev/null | ssh  -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null  -i /run/secrets/zfs/id_ed25519 -S /tmp/syncoid-zfs@backuphost-1628463600 zfs@backuphost ' mbuffer  -q -s 128k -m 16M 2>/dev/null | lzop -dfc |  zfs receive  -s -F '"'"'bpool/backup/myhost/safe'"'"' 2>&1' failed: 256 at /nix/store/ygx9z7xqky03cvymfsja24xmmf8yd7hy-sanoid-2.1.0/bin/.syncoid-wrapped line 817.
Aug 09 02:00:31 myhost systemd[1]: syncoid-rpool-safe.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Aug 09 02:00:31 myhost systemd[1]: syncoid-rpool-safe.service: Failed with result 'exit-code'.
Aug 09 02:00:31 myhost systemd[1]: syncoid-rpool-safe.service: Consumed 10.039s CPU time, received 5.9M IP traffic, sent 448.6M IP traffic.
```

Here is my config:
```nix
{
  services.sanoid = {
    enable = true;
    interval = "daily";
    datasets = {
      "rpool/safe" = {
        autosnap = true;
        autoprune = true;
        recursive = true;
        hourly = 0;
        daily = 31;
        monthly = 3;
        yearly = 0;
      };
    };
  };

  services.syncoid = {
    enable = mkDefault true;
    interval = "02:00";
    commonArgs = [
      "--no-sync-snap"
      "--create-bookmark"
    ];
    commands = {
      "rpool/safe" = {
        sshKey = config.age.secrets."zfs/id_ed25519".path;
        extraArgs = [
          "--sshoption=StrictHostKeyChecking=no"
          "--sshoption=UserKnownHostsFile=/dev/null"
        ];
        sendOptions = "-wR";
        target = mkDefault "zfs@backuphost:bpool/backup/${config.networking.hostName}/safe";
      };
    };
  };
}
```

Is there something I'm doing wrong?

etu commented:
Seems fine here:
```
Aug 08 19:15:00 agrajag systemd[1]: Starting Syncoid ZFS synchronization from zroot/home to root@home.elis.nu:zroot/backups/agrajag/zroot/home...
Aug 08 19:15:00 agrajag systemd[1]: Started Syncoid ZFS synchronization from zroot/home to root@home.elis.nu:zroot/backups/agrajag/zroot/home.
Aug 08 19:15:01 agrajag syncoid[118019]: WARN: lzop not available on target ssh:-S /tmp/syncoid-root@home.elis.nu-1628442901 root@home.elis.nu - sync will continue without compression.
Aug 08 19:15:01 agrajag syncoid[118019]: WARN: mbuffer not available on target ssh:-S /tmp/syncoid-root@home.elis.nu-1628442901 root@home.elis.nu - sync will continue without target buffering.
Aug 08 19:15:02 agrajag syncoid[118019]: NEWEST SNAPSHOT: autosnap_2021-08-08_17:15:01_frequently
Aug 08 19:15:02 agrajag syncoid[118019]: Sending incremental zroot/home@autosnap_2021-08-08_16:15:07_frequently ... autosnap_2021-08-08_17:15:01_frequently (~ 76.0 MB):
Aug 08 19:15:11 agrajag systemd[1]: syncoid-zroot-home.service: Succeeded.
Aug 08 19:15:11 agrajag systemd[1]: syncoid-zroot-home.service: Consumed 1.362s CPU time, received 12.0M IP traffic, sent 75.0M IP traffic.
Aug 08 20:15:03 agrajag systemd[1]: Starting Syncoid ZFS synchronization from zroot/home to root@home.elis.nu:zroot/backups/agrajag/zroot/home...
Aug 08 20:15:03 agrajag systemd[1]: Started Syncoid ZFS synchronization from zroot/home to root@home.elis.nu:zroot/backups/agrajag/zroot/home.
Aug 08 20:15:04 agrajag syncoid[179514]: WARN: lzop not available on target ssh:-S /tmp/syncoid-root@home.elis.nu-1628446503 root@home.elis.nu - sync will continue without compression.
Aug 08 20:15:04 agrajag syncoid[179514]: WARN: mbuffer not available on target ssh:-S /tmp/syncoid-root@home.elis.nu-1628446503 root@home.elis.nu - sync will continue without target buffering.
Aug 08 20:15:05 agrajag syncoid[179514]: NEWEST SNAPSHOT: autosnap_2021-08-08_18:15:03_frequently
Aug 08 20:15:05 agrajag syncoid[179514]: Sending incremental zroot/home@autosnap_2021-08-08_17:15:01_frequently ... autosnap_2021-08-08_18:15:03_frequently (~ 273.3 MB):
Aug 08 20:15:34 agrajag systemd[1]: syncoid-zroot-home.service: Succeeded.
Aug 08 20:15:34 agrajag systemd[1]: syncoid-zroot-home.service: Consumed 3.661s CPU time, received 41.3M IP traffic, sent 271.9M IP traffic.

[cut a bunch of hours of logs that just looked the same]

Aug 09 20:15:24 agrajag systemd[1]: Starting Syncoid ZFS synchronization from zroot/home to root@home.elis.nu:zroot/backups/agrajag/zroot/home...
Aug 09 20:15:24 agrajag systemd[1]: Started Syncoid ZFS synchronization from zroot/home to root@home.elis.nu:zroot/backups/agrajag/zroot/home.
Aug 09 20:15:25 agrajag syncoid[77167]: WARN: lzop not available on target ssh:-S /tmp/syncoid-root@home.elis.nu-1628532924 root@home.elis.nu - sync will continue without compression.
Aug 09 20:15:25 agrajag syncoid[77167]: WARN: mbuffer not available on target ssh:-S /tmp/syncoid-root@home.elis.nu-1628532924 root@home.elis.nu - sync will continue without target buffering.
Aug 09 20:15:26 agrajag syncoid[77167]: NEWEST SNAPSHOT: autosnap_2021-08-09_18:15:24_frequently
Aug 09 20:15:26 agrajag syncoid[77167]: Sending incremental zroot/home@autosnap_2021-08-09_08:15:38_frequently ... autosnap_2021-08-09_18:15:24_frequently (~ 217.5 MB):
Aug 09 20:15:49 agrajag systemd[1]: syncoid-zroot-home.service: Succeeded.
Aug 09 20:15:49 agrajag systemd[1]: syncoid-zroot-home.service: Consumed 3.413s CPU time, received 32.5M IP traffic, sent 211.6M IP traffic.
```

It for sure **looks** like it only sends a single snapshot, but here's a list of my 20 last snapshots on the system where I took these logs from:
```
zroot/home@autosnap_2021-08-08_15:00:07_hourly      5.98M      -     18.3G  -
zroot/home@autosnap_2021-08-08_16:00:07_hourly      5.79M      -     18.3G  -
zroot/home@autosnap_2021-08-08_17:00:07_hourly      5.74M      -     18.3G  -
zroot/home@autosnap_2021-08-08_18:00:06_hourly      27.3M      -     18.4G  -
zroot/home@autosnap_2021-08-08_19:00:02_hourly      30.8M      -     18.5G  -
zroot/home@autosnap_2021-08-08_20:00:05_hourly      13.7M      -     18.5G  -
zroot/home@autosnap_2021-08-09_06:15:06_daily          0B      -     18.5G  -
zroot/home@autosnap_2021-08-09_06:15:06_hourly         0B      -     18.5G  -
zroot/home@autosnap_2021-08-09_07:00:06_hourly      9.98M      -     18.5G  -
zroot/home@autosnap_2021-08-09_08:00:38_hourly      20.2M      -     18.3G  -
zroot/home@autosnap_2021-08-09_08:15:38_frequently  19.5M      -     18.4G  -
zroot/home@autosnap_2021-08-09_08:30:00_frequently  29.7M      -     18.4G  -
zroot/home@autosnap_2021-08-09_08:45:39_frequently  16.6M      -     14.8G  -
zroot/home@autosnap_2021-08-09_17:51:36_hourly         0B      -     14.8G  -
zroot/home@autosnap_2021-08-09_17:51:36_frequently     0B      -     14.8G  -
zroot/home@autosnap_2021-08-09_18:00:39_hourly         0B      -     14.8G  -
zroot/home@autosnap_2021-08-09_18:00:39_frequently     0B      -     14.8G  -
zroot/home@autosnap_2021-08-09_18:15:24_frequently  26.9M      -     14.8G  -
zroot/home@autosnap_2021-08-09_18:30:39_frequently  21.2M      -     14.8G  -
zroot/home@autosnap_2021-08-09_18:45:24_frequently  9.82M      -     14.8G  -
```

And this is the 20 last snapshots of where I send them to:
```
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_14:00:07_hourly      20.9M      -     18.3G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_15:00:07_hourly      5.95M      -     18.3G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_16:00:07_hourly      5.75M      -     18.3G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_17:00:07_hourly      5.70M      -     18.3G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_18:00:06_hourly      27.2M      -     18.4G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_19:00:02_hourly      30.8M      -     18.5G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-08_20:00:05_hourly      13.6M      -     18.5G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_06:15:06_daily          0B      -     18.5G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_06:15:06_hourly         0B      -     18.5G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_07:00:06_hourly      9.94M      -     18.5G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_08:00:38_hourly         0B      -     18.3G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_08:00:38_frequently     0B      -     18.3G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_08:15:38_frequently  19.4M      -     18.4G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_08:30:00_frequently  29.6M      -     18.4G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_08:45:39_frequently  16.6M      -     14.8G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_17:51:36_hourly         0B      -     14.8G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_17:51:36_frequently     0B      -     14.8G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_18:00:39_hourly         0B      -     14.8G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_18:00:39_frequently     0B      -     14.8G  -
zroot/backups/agrajag/zroot/home@autosnap_2021-08-09_18:15:24_frequently     0B      -     14.8G  -
```

So of course I miss the last two frequently snapshots on the target host since time is 18:50 UTC as of writing, and last sync action was 18:15 UTC. So snapshots 18:30 UTC and 18:45 UTC has been created locally but not sent of yet since I sync at `*:15:00`. But as you can see it sends more than one snapshot. Then I have sanoid cleaning up certain old snapshots on the other side to match the retention rules that I use on my laptop.

### Actually looking at your data:
I find the end of this interesting:
```
Aug 09 02:00:31 myhost systemd[1]: syncoid-rpool-safe.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

I also find it a bit odd to have the options to ignore the known hosts file rather than making sure it's the correct host.

It just dawned on me, I'm pretty sure that the issue isn't what's done in this PR, the issue you're facing is from this previous PR that just got merged #98455. I think it's due to the new hardening of the syncoid services. Looking at your configuration, you keep your secret ssh key in agenix which make them show up in `/var/run/secrets`, which I'm fairly certain that the syncoid service aren't allowed to read from.

What I would do is to add the following:
```nix
{
  # Bind mount syncoid ssh key
  fileSystems."/var/lib/syncoid/.ssh/id_ed25519" = {
    device = config.age.secrets."zfs/id_ed25519".path;
    options = [ "bind" "noauto" "x-systemd.automount" ];
  };
}
```

misuzu commented:
> which I'm fairly certain that the syncoid service aren't allowed to read from.

No, it's fine. It sends exactly ONE snapshot and then fails.

misuzu commented:
I've tried reverting all changes including #98455, but it still failed with the same `status=2/INVALIDARGUMENT` error. I guess I'll just try something else instead of sanoid/syncoid.

etu commented:
@misuzu Did you revert the changes from this PR to the syncoid module as well? Just to make sure.

misuzu commented:
> @misuzu Did you revert the changes from this PR to the syncoid module as well? Just to make sure.

Yes. The issue appears to be with `zfs receive` exiting with exit code `1` on success when started under unprivileged user. I don't know what's going on, since there are no error messages. I had to switch to `root` account to make this go away.

etu commented:
@misuzu I'm thinking this may be some race-condition between syncoid and sanoid, say that you run sanoid as the same user on the target system as syncoid is connecting to. Then that the sanoid service delegates zfs permissions for sanoid to work which then is reverted by the service after sanoid is done. If you, connected to this user to send snapshots there, that user would have extra permissions for a bit of time.

misuzu commented:
> @misuzu I'm thinking this may be some race-condition between syncoid and sanoid, say that you run sanoid as the same user on the target system as syncoid is connecting to. Then that the sanoid service delegates zfs permissions for sanoid to work which then is reverted by the service after sanoid is done. If you, connected to this user to send snapshots there, that user would have extra permissions for a bit of time.

This is something different. I've stored output of `zfs send` to a file, transferred this file to the remote server and tried `zfs receive` on it using unprivileged account and `root`. `zfs receive` doing its thing just fine in both cases, but exits with exit code `1` on unprivileged account. There aren't any error messages even with `-v` option.

ju1m commented:
@misuzu, indeed it looks like this 2015 bug https://github.com/openzfs/zfs/issues/3479 (closed by inactivity, not by a fix). But I cannot reproduce it. Putting clones in the mix seems [prone to bug](https://stackoverflow.com/a/65627689). But this is out of my knowledge zone for me to say. This said, the fact that your still getting this error by invoking `zfs` manually let me think that it's not related to NixOS' `syncoid` module.

Here's what I've tried to reproduce your exit code `1`, but to no avail:
```
#!/bin/sh -eux
origin=oignon/test-origin
clones=oignon/test-clones
backup=off2/julm/backup/test-clones

zfs destroy -R $origin || true
zfs destroy -R $clones || true
zfs destroy -R $backup || true

zfs create $origin
zfs create -V 1G $origin/d0
zfs snapshot $origin/d0@snap

zfs create $clones
zfs clone $origin/d0@snap $clones/d0
zfs clone $origin/d0@snap $clones/d1
zfs promote $clones/d0
zfs destroy -r $origin/d0
#zfs create -V 1G $clones/d2

zfs snapshot -r $clones@snap1
zfs send -vwR $clones@snap1 | zfs recv -uvF $backup
echo errno=$?

zfs snapshot -r $clones@snap2
zfs send -vwR -i $clones@snap1 $clones@snap2 | zfs recv -uvF $backup
echo errno=$?

zfs list -t all -o name,origin,clones | grep test-clones
```

```
+ origin=oignon/test-origin
+ clones=oignon/test-clones
+ backup=off2/julm/backup/test-clones
+ zfs destroy -R oignon/test-origin
+ zfs destroy -R oignon/test-clones
+ zfs destroy -R off2/julm/backup/test-clones
+ zfs create oignon/test-origin
+ zfs create -V 1G oignon/test-origin/d0
+ zfs snapshot oignon/test-origin/d0@snap
+ zfs create oignon/test-clones
+ zfs clone oignon/test-origin/d0@snap oignon/test-clones/d0
+ zfs clone oignon/test-origin/d0@snap oignon/test-clones/d1
+ zfs promote oignon/test-clones/d0
+ zfs destroy -r oignon/test-origin/d0
+ zfs snapshot -r oignon/test-clones@snap1
+ zfs recv -uvF off2/julm/backup/test-clones
+ zfs send -vwR oignon/test-clones@snap1
full send of oignon/test-clones@snap1 estimated size is 69.6K
full send of oignon/test-clones/d0@snap estimated size is 42.0K
send from @snap to oignon/test-clones/d0@snap1 estimated size is 624B
full send of oignon/test-clones/d1@snap1 estimated size is 42.0K
total estimated size is 154K
receiving full stream of oignon/test-clones@snap1 into off2/julm/backup/test-clones@snap1
received 22.2K stream in 1 seconds (22.2K/sec)
receiving full stream of oignon/test-clones/d0@snap into off2/julm/backup/test-clones/d0@snap
received 4.55K stream in 1 seconds (4.55K/sec)
receiving incremental stream of oignon/test-clones/d0@snap1 into off2/julm/backup/test-clones/d0@snap1
received 1.31K stream in 1 seconds (1.31K/sec)
found clone origin off2/julm/backup/test-clones/d0@snap
receiving incremental stream of oignon/test-clones/d1@snap1 into off2/julm/backup/test-clones/d1@snap1
received 1.31K stream in 1 seconds (1.31K/sec)
+ echo errno=0
errno=0
+ zfs snapshot -r oignon/test-clones@snap2
+ zfs send -vwR -i oignon/test-clones@snap1 oignon/test-clones@snap2
+ zfs recv -uvF off2/julm/backup/test-clones
send from @snap1 to oignon/test-clones@snap2 estimated size is 624B
send from @snap1 to oignon/test-clones/d0@snap2 estimated size is 624B
send from @snap1 to oignon/test-clones/d1@snap2 estimated size is 624B
total estimated size is 1.83K
receiving incremental stream of oignon/test-clones@snap2 into off2/julm/backup/test-clones@snap2
received 1.31K stream in 1 seconds (1.31K/sec)
receiving incremental stream of oignon/test-clones/d0@snap2 into off2/julm/backup/test-clones/d0@snap2
received 1.31K stream in 1 seconds (1.31K/sec)
receiving incremental stream of oignon/test-clones/d1@snap2 into off2/julm/backup/test-clones/d1@snap2
received 1.31K stream in 1 seconds (1.31K/sec)
+ echo errno=0
errno=0
+ zfs list -t all -o name,origin,clones
+ grep test-clones
off2/julm/backup/test-clones                                                   -                                     -
off2/julm/backup/test-clones@snap1                                             -                                     -
off2/julm/backup/test-clones@snap2                                             -                                     -
off2/julm/backup/test-clones/d0                                                -                                     -
off2/julm/backup/test-clones/d0@snap                                           -                                     off2/julm/backup/test-clones/d1
off2/julm/backup/test-clones/d0@snap1                                          -                                     -
off2/julm/backup/test-clones/d0@snap2                                          -                                     -
off2/julm/backup/test-clones/d1                                                off2/julm/backup/test-clones/d0@snap  -
off2/julm/backup/test-clones/d1@snap1                                          -                                     -
off2/julm/backup/test-clones/d1@snap2                                          -                                     -
oignon/test-clones                                                             -                                     -
oignon/test-clones@snap1                                                       -                                     -
oignon/test-clones@snap2                                                       -                                     -
oignon/test-clones/d0                                                          -                                     -
oignon/test-clones/d0@snap                                                     -                                     oignon/test-clones/d1
oignon/test-clones/d0@snap1                                                    -                                     -
oignon/test-clones/d0@snap2                                                    -                                     -
oignon/test-clones/d1                                                          oignon/test-clones/d0@snap            -
oignon/test-clones/d1@snap1                                                    -                                     -
oignon/test-clones/d1@snap2                                                    -                                     -
```

misuzu commented:
@ju1m my setup looks something like this:

```bash
#!/usr/bin/env bash
sudo zpool destroy -f opool || true
sudo zpool destroy -f bpool || true

sudo fallocate -l 128M /tmp/opool.zfs
sudo fallocate -l 128M /tmp/bpool.zfs

echo -e "passwd12345\npasswd12345" | sudo zpool create -f \
    -O atime=off -O xattr=sa -O acltype=posixacl \
    -O encryption=on \
    -O keyformat=passphrase \
    -O mountpoint=none \
    opool \
    /tmp/opool.zfs

sudo zpool create -f \
    -O atime=off -O xattr=sa -O acltype=posixacl \
    -O mountpoint=none \
    bpool \
    /tmp/bpool.zfs

origin=opool/safe
backup=bpool/backup

sudo zfs destroy -R $origin || true
sudo zfs destroy -R $backup || true

sudo zfs create -o mountpoint=none $origin
sudo zfs create -o mountpoint=legacy $origin/root
sudo zfs snapshot -r $origin@snap

sudo zfs create -o mountpoint=none $backup

sudo zfs allow $(whoami) compression,create,destroy,mount,mountpoint,receive,rollback,send,snapshot,bookmark,hold $backup

sudo zfs send -vwR $origin@snap | zfs recv -uvsF $backup/myhost
echo errno=$?

zfs list -t all -o name | grep $backup
```

```
$ sudo bash ./test.sh
cannot open 'opool/safe': dataset does not exist
cannot open 'bpool/backup': dataset does not exist
full send of opool/safe@snap estimated size is 49.6K
full send of opool/safe/root@snap estimated size is 49.6K
total estimated size is 99.2K
receiving full stream of opool/safe@snap into bpool/backup/myhost@snap
received 16.7K stream in 1 seconds (16.7K/sec)
receiving full stream of opool/safe/root@snap into bpool/backup/myhost/root@snap
received 16.7K stream in 1 seconds (16.7K/sec)
errno=0
bpool/backup
bpool/backup/myhost
bpool/backup/myhost@snap
bpool/backup/myhost/root
bpool/backup/myhost/root@snap
$ bash ./test.sh
cannot open 'opool/safe': dataset does not exist
cannot open 'bpool/backup': dataset does not exist
full send of opool/safe@snap estimated size is 49.6K
full send of opool/safe/root@snap estimated size is 49.6K
total estimated size is 99.2K
receiving full stream of opool/safe@snap into bpool/backup/myhost@snap
received 16.7K stream in 1 seconds (16.7K/sec)
receiving full stream of opool/safe/root@snap into bpool/backup/myhost/root@snap
received 16.7K stream in 1 seconds (16.7K/sec)
errno=1
bpool/backup
bpool/backup/myhost
bpool/backup/myhost@snap
bpool/backup/myhost/root
bpool/backup/myhost/root@snap
```

ju1m commented:
@misuzu I can reproduce. Using `sudo zfs recv` exits successfully. And after bisecting on the missing `zfs allow` permissions, just adding `change-key` to the list also exits successfully for me :)

ju1m commented:
@etu, there is a regression due to delegating the permissions only to the datasets used.
When the receiving dataset does not exist yet, setting the permission fails, and then `syncoid` cannot create the dataset.

One is not supposed to create the dataset manually, that would fail with:
```
Started Syncoid ZFS synchronization from backup@mermet.sourcephile.fr:rpool/var/postgresql to losurdo/backup/mermet/var/postgresql.
NEWEST SNAPSHOT: autosnap_2021-08-29_13:00:12_hourly                                
CRITICAL ERROR: Target losurdo/backup/mermet/var/postgresql exists but has no snapshots matching with rpool/var/postgresql!
Replication to target would require destroying existing             
target. Cowardly refusing to destroy your existing target.          
NOTE: Target losurdo/backup/mermet/var/postgresql dataset is < 64MB used - did you mistakenly run
`zfs create losurdo/backup/mermet/var/postgresql` on the target? ZFS initial
replication must be to a NON EXISTENT DATASET, which will           
then be CREATED BY the initial replication process.
```
A workaround is to manually apply permissions: `zfs allow syncoid create,mount,receive` to the parent dataset.
So I guess the `ExecStartPre=` should try to set "mount,receive,rollback" on the receiving dataset if it exists, or try to set `create,mount,receive` on the parent dataset.

etu commented:
@ju1m I've been putting some thought into this, and yeah, you're right, this is a problem and I think we should solve it, especially before the next release.

I see two ways, as you mention, we would need to apply permissions to the parent dataset. This is not something we could determine in nix, it has to be in runtime. Especially since it will change from first to second run.

The other way would be to document this within the `configuration.nix` file, but I'd rather solve it in the unit to be done automatically.

SemMulder commented:
I'm trying this out, but I'm running into the following:
```
Oct 15 18:00:09 meterkassie systemd[1]: Started Sanoid snapshot service.
Oct 15 18:00:10 meterkassie sanoid[4644]: Permission denied the ZFS utilities must be run as root.
Oct 15 18:00:10 meterkassie sanoid[4645]: Permission denied the ZFS utilities must be run as root.
Oct 15 18:00:10 meterkassie sanoid[4639]: CRITICAL ERROR: zfs snapshot state/photos@autosnap_2021-10-15_18:00:10_monthly failed, 256 at /nix/>
Oct 15 18:00:10 meterkassie sanoid[4647]: Permission denied the ZFS utilities must be run as root.
```

Which to me seems like the ExecStartPre lines are not running as root? Even though they are prefixed with `+`. Anybody ran into this before?
